File: vokimi_frontend/src/ts/page_classes/manage_test_page/test_statistics/templates/general/GeneralTestStatisticsData.ts
import type { ITestStatisticsData } from "../../ITestStatisticsData";
import type { TestStatisticsDiscussionsData } from "../../templates_shared/TestStatisticsDiscussionsData";
import type { TestStatisticsRatingsData } from "../../templates_shared/TestStatisticsRatingsData";
import type { TestStatisticsTestTakenRecordsCount } from "../../templates_shared/TestStatisticsTestTakenRecordsCount";

export class GeneralTestStatisticsData implements ITestStatisticsData {
    testName: string;
    testTakenRecords: TestStatisticsTestTakenRecordsCount;
    ratings: TestStatisticsRatingsData;
    discussions: TestStatisticsDiscussionsData;
    resultsData: GeneralTestStatisticsData[];

    mostPopularResult(): GeneralTestStatisticsData {
        return this.resultsData.reduce((previous, current) => {
            return previous.testTakenRecords.all > current.testTakenRecords.all ? previous : current;
        }, this.resultsData[0]);
    }
    constructor(
        testName: string,
        testTakenRecords: TestStatisticsTestTakenRecordsCount,
        ratings: TestStatisticsRatingsData,
        discussions: TestStatisticsDiscussionsData,
        resultsData: GeneralTestStatisticsData[]
    ) {
        this.testName = testName;
        this.testTakenRecords = testTakenRecords;
        this.ratings = ratings;
        this.discussions = discussions;
        this.resultsData = resultsData;
    }
}

File: vokimi_api/Src/dtos/responses/manage_test_page/test_statistics/templates/general/GeneralTestStatisticsData.cs
using vokimi_api.Src.dtos.responses.manage_test_page.test_statistics.templates_shared;

namespace vokimi_api.Src.dtos.responses.manage_test_page.test_statistics.templates.general
{
    public record GeneralTestStatisticsData(
        TestStatisticsTestTakenRecordsCount TestTakenRecords,
        TestStatisticsRatingsData Ratings,
        TestStatisticsDiscussionsData Discussions
        ) : BaseTestStatisticsData(TestTakenRecords, Ratings, Discussions)
    {

    }
}
